src/careamics/dataset/tiling/lvae_tiled_patching.py

```diff
@@ -0,0 +1,282 @@
+    tile_size: NDArray[np.int_],
+    overlaps: NDArray[np.int_],
+) -> tuple[tuple[int, int], ...]:
+    """
+    Calculate padding to ensure stitched data comes from the center of a tile.
+
+    Padding is added to an array with shape `data_shape` so that when tiles are
+    stitched together, the data used always comes from the center of a tile, even for
+    tiles at the boundaries of the array.
+
+    Parameters
+    ----------
+    data_shape : 1D numpy.array of int
+        The shape of the data to be tiled and stitched together, (S, C, (Z), Y, X).
+    tile_size : 1D numpy.array of int
+        The tile size in each dimension, ((Z), Y, X).
+    overlaps : 1D numpy.array of int
+        The tile overlap in each dimension, ((Z), Y, X).
+
+    Returns
+    -------
+    tuple of (int, int)
+        A tuple specifying the padding to add in each dimension, each element is a two
+        element tuple specifying the padding to add before and after the data. This
+        can be used as the `pad_width` argument to `numpy.pad`.
+    """
+    tile_grid_shape = np.array(compute_tile_grid_shape(data_shape, tile_size, overlaps))
+    covered_shape = (tile_size - overlaps) * tile_grid_shape + overlaps
+
+    pad_before = overlaps // 2
+    pad_after = covered_shape - data_shape[-len(tile_size) :] - pad_before
+
+    return tuple((before, after) for before, after in zip(pad_before, pad_after))
+
+
+def n_tiles_1d(axis_size: int, tile_size: int, overlap: int) -> int:
+    """Calculate the number of tiles in a specific dimension.
+
+    Parameters
+    ----------
+    axis_size : int
+        The length of the data for in a specific dimension.
+    tile_size : int
+        The length of the tiles in a specific dimension.
+    overlap : int
+        The tile overlap in a specific dimension.
+
+    Returns
+    -------
+    int
+        The number of tiles that fit in one dimension given the arguments.
+    """
+    return int(np.ceil(axis_size / (tile_size - overlap)))
+
+
+def total_n_tiles(
+    data_shape: tuple[int, ...], tile_size: tuple[int, ...], overlaps: tuple[int, ...]
+) -> int:
+    """Calculate The total number of tiles over all dimensions.
+
+    Parameters
+    ----------
+    data_shape : 1D numpy.array of int
+        The shape of the data to be tiled and stitched together, (S, C, (Z), Y, X).
+    tile_size : 1D numpy.array of int
+        The tile size in each dimension, ((Z), Y, X).
+    overlaps : 1D numpy.array of int
+        The tile overlap in each dimension, ((Z), Y, X).
+
+
+    Returns
+    -------
+    int
+        The total number of tiles over all dimensions.
+    """
+    result = 1
+    # assume spatial dimension are the last dimensions so iterate backwards
+    for i in range(-1, -len(tile_size) - 1, -1):
+        result = result * n_tiles_1d(data_shape[i], tile_size[i], overlaps[i])
+
+    return result
+
+
+def compute_tile_grid_shape(
+    data_shape: NDArray[np.int_],
+    tile_size: NDArray[np.int_],
```

Comment:
overlap  will be a tuple in 3D. So, let's say the model takes as input 5x64x64, where 5 is the z-dimension. so, in your notation this is the tile size. However, we could have`4x48x48` as the overlap so that only the `16x16` of the central z slice in the prediction will be kept and 2 slices from the beginning and the end will be removed. But nothing to be done for now. just writing here to make you aware.

Reply:
So some of the boundary z-slices are discarded during prediction?

Reply:
I had a quick look at the changes in `GridIndexManager`. It is my understanding that the tiles are shifted so that they are centred on the z-axis, and tiles are not permitted to exceed the bounds of the image (to remove the need for padding). But, in most cases (if image not equally divisible by the grid size), this necessarily leads to some of the boundary z-slices being discarded. Is this correct?

N.b. This implementation handles tiling the exact same way as yours, I just reformulated it so I would be able to integrate it with CAREamics prediction pipeline more easily. The boundary tiles are padded the same, and then the padding is removed during `stitch_prediction`. The area to stitch together is specified by `overlap_crop_coords` in `TileInformation`; the calculation of `overlap_crop_coords` in `compute_tile_info` does not allow beyond the bounds of the original image to be included.

Reply:
I guess we can discuss face to face. Whenever you are in the lab. It might be more efficient that way. Since you have changed the structure, it can take me sometime to fully understand which I would prefer to not give :) . Face to face would be better. 

No z-slices are discarded in my current version. And I have overlap in x,y and z.

Reply:
I see, I will have a proper look at the changes in `GridIndexManager` because I have only skimmed it, and I'll try to get a better understanding. I think I can re-work this so that the tiling strategy can be specified per dimension, but it will basically require me to rewrite the entire thing.

I will be in the lab tomorrow, if you're available?